### PR TITLE
Give AWS object client the ability to get creds from vault

### PR DESCRIPTION
**Pull request by msteffen**

### CODE REVIEW

File: src/server/pkg/obj/amazon_client.go [hunk 1]
```diff
@@ -37,16 +44,143 @@ type amazonClient struct {
...
+		return emptyCreds, fmt.Errorf("aws creds in vault response were not both strings (%T and %T)", accessKeyIface, awsSecretIface)
+	}
+
+	// update the lease values in 'v', and spawn a goroutine to renew the lease
+	v.updateLease(vaultSecret)
+	go func() {
+		for {
+			// renew at half the lease duration or one day, whichever is greater
+			// (lease must expire eventually)
+			renewInterval := v.leaseDuration
```
Comment (jdoliner) [inline src/server/pkg/obj/amazon_client.go]: Should this be holding the mutex will it accesses v.leaseDuration? It looks like that could possibly race with the code above. But I'm not quite sure.

File: src/server/pkg/obj/obj.go [hunk 2]
```diff
@@ -360,12 +375,13 @@ func (b *BackoffWriteCloser) Write(data []byte) (int, error) {
 			return err
 		}
 		return nil
-	}, b.backoffConfig, func(err error, d time.Duration) {
+	}, b.backoffConfig, func(err error, d time.Duration) error {
```
Comment (jdoliner) [inline src/server/pkg/obj/obj.go]: Weird... how was this compiling before?

Comment (msteffen) [reply src/server/pkg/obj/obj.go]: It was using a different library. One of the changes in this PR is that I replaced `github.com/cenkalti/backoff` with `github.com/pachyderm/pachyderm/src/server/pkg/backoff`

File: src/server/pkg/obj/amazon_client.go (hunk 1 above)
Comment (msteffen) [reply src/server/pkg/obj/amazon_client.go]: Good catch. I'll rewrite this code a bit

File: src/server/pkg/obj/obj.go (hunk 2 above)
Comment (jdoliner) [reply src/server/pkg/obj/obj.go]: Ahh, right. Makes sense.

**Pull request merged**